rfc-149-switch-to-per-page-asset-loading.md

```diff
@@ -0,0 +1,208 @@
+  # check:
+  #  - if the component has already had the assets added to the page
+  # - if the component assets loading method is allowed at an application level
+  # - if the individual component has used an option to turn individual assets on or off
+  @components_in_use.include?(component_name) || ENV.component_asset_loading !== "bulk" || use_individual_assets === true
+end
+```
+
+## Other solutions that have been considered and rejected
+
+* Put all components used across GOV.UK in one stylesheet and one JavaScript file - this increases page weight and doesn't solve cache invalidation
+* Serve one stylesheet and one JavaScript file for all of GOV.UK - this increases page weight and doesn't solve cache invalidation
+* Serve one stylesheet and one JavaScript file per template - this doesn't solve cache invalidation when one component changes
+* Serve all component assets used in an application in individual files - this increases page weight for single page visitors, though often decreased it for people who visited more than one page
+
+## Implementation process
+
+Serving assets individually doesn't rely on the shared asset folder proposed in RFC 91 - so the lack of a shared folder is not a blocker.
+
+* implement the individual serving of assets
+   1. Stylesheets
+   1. JavaScript, as this will require more investigation and relies on external factors
+* once a shared folder is available, make all applications load assets from the shared folder
+
+### Implement the individual serving of assets
+
+Splitting this into two stages makes it more manageable and will allow improvements to be in stages.
+
+The stages themselves can be split into app-by-app upgrades - once an app is upgraded the pages it renders will see performance improvements without relying on any other app also being upgraded. This is a similar approach to the work that was done to use the public layout component, so we know it can work - even if it takes time.
+
+This shouldn't be done on a component by component basis - changing only one component on a page will likely cause an increase in page size as the individual component assets can't take advantage of compression.
+
+There are more unknowns with the JavaScript implementation - and this aspect of the work may require further investigation and changes to applications.
+
+It's worth noting that ES6 modules and import maps are the [direction of travel that Ruby on Rails is taking][rails_direction], so this could be the path of least resistance that this implementation could take. GOV.UK Frontend also offers module based imports now, and has plans to drop JavaScript support for Internet Explorer 11 in the near future. It'd be worth seeing how close to a decision these are before choosing a direction.
+
+### Make assets load from a shared folder
+
+Parallel to this, we need to see how RFC 91 could be implemented. This is complicated by the replatforming project nearing completion.
```

Comment:
Is it? I'm not sure that replatforming will change the asset compilation/upload pipeline 🤔 I think any changes we make in this area would be carried over automatically - though @sengi will be able to confirm.

Reply:
Because RFC 91 hasn't been implemented there is no shared folder that shared assets could be uploaded to.

We could implement a shared folder for the current set up - but from discussions it seems likely that this would be finished just in time for replatforming to be switched on.

Possibly complicated is the wrong word - I'm try to note that the shared folder is an external factor with its own RFC, and finding out how this should be carried out is a piece of work in itself.

Reply:
In the new system, Rails assets will be uploaded to a common S3 bucket on rollout (i.e. when a version of an app is deployed to the cluster). It should be feasible to share assets. Right now we're still serving files via the cluster instead of going straight to S3 from the edge (Fastly), but we'd like to do that if not before launch then soon after.

Reply:
https://docs.google.com/drawings/d/1UjW5El9AnrqzgdXjAXKfYBNFR0xNth54hT1PvnMsPxc/edit

Reply:
We should have a more concrete time frame soon but we're talking this year.

Reply:
That's super news @sengi! Apologies, I was unaware that a shared folder was being implemented so I incorrectly assumed that it wasn't part of the new system.

I'll update this RFC to add that a shared folder is coming soon, and what this will mean for this proposal.

Reply:
I've updated it to make it clearer that the shared folder is coming soon - https://github.com/alphagov/govuk-rfcs/pull/152/commits/09a5bd268a01f73da3ab6b56da309b03652a5497. Does this sound correct?